neuro_evolution_ctrnn/brain_visualizer/__init__.py

```diff
@@ -1,40 +1,260 @@
+
+
 
     def process_update(self, in_values, out_values):
-        self.app.set_text("new states: \n" + str(self.brain.y))
-        self.root.update_idletasks()
-        self.root.update()
-
-
-class Window(tk.Frame):
-    def __init__(self, master=None):
-        tk.Frame.__init__(self, master)
-        self.master = master
-        self.pack(fill=tk.BOTH, expand=1)
-        self.text = tk.StringVar()
-        self.set_text("waiting for update...")
-        tk.Label(self, textvariable=self.text).pack()
-
-    def set_text(self, text):
-        self.text.set(text)
+        # Fill screen with color
+        self.screen.fill((self.displayColor))
+
+        # Draw Rect for Logo and Blit Logo on the screen
+        pygame.draw.rect(self.screen, self.darkGrey, (0, 0, self.w, 60))
+        self.screen.blit(self.neuroboticsLogo, self.neuroboticsRect)
+        self.screen.blit(self.kitLogo, self.kitRect)
+
+        ##### Number Neurons
+        numberInputNeurons = len(in_values)
+        numberNeurons = len(self.brain.W.todense())
+        numberOutputNeurons = len(out_values)
+
+        ########## Draw Legend
+        PygameBrainVisualizer.render_InfoText(self, [
+            "Input Neurons: " + str(numberInputNeurons),
+            "Graph Neurons: " + str(numberNeurons),
+            "Output Neurons: " + str(numberOutputNeurons)],
+                                              ((self.w / 4) - 80), 5, self.numColor, 18)
+
+        PygameBrainVisualizer.render_InfoText(self, [
+            "Positive/Negative Weights [t,w] : " + str(self.positiveWeights) + " / " + str(self.negativeWeights),
+            "Input/Output Weights [q,z] : " + str(self.inputWeights) + " / " + str(self.outputWeights),
+            "Direction [g] : " + str(self.weightsDirection)],
+                        ((self.w / 2) - 130), 5, self.numColor, 18)
+
+        if self.weightVal == 0:     # If weightVal is 0 every Connection will be drawn
+            text = "all"
+        else:
+            text = str(self.weightVal)
+        PygameBrainVisualizer.render_InfoText(self, [
+            "Weights [e,r] : " + text,
+            "Values [s] : " + str(self.neuronText),
+            "Simulation : " + str(self.environment)],
+                        ((3 * self.w / 4) - 80), 5, self.numColor, 18)
+
+
+
+        ########## Create Dictionaries with Positions
+        ##### Input Dictionary
+        inputPositionsDict = Positions.getInputOutputPositions(self, numberInputNeurons, True)
+
+        ##### Dictionary Graph Neurons
+        # --> self.graphPositionsDict
+
+        ##### Output Dictionary
+        outputPositionsDict = Positions.getInputOutputPositions(self, numberOutputNeurons, False)
+
+        ########## Draw Weights
+        ##### n-1 Linien pro Neuron ; Input zu Neuron
+        if self.inputWeights:
+            Weights.draw(self, inputPositionsDict, self.graphPositionsDict, self.brain.V.todense().T, self.positiveWeights, self.negativeWeights, self.weightsDirection)
```

Comment:
Der Methodenaufruf bringt das Programm bei mir zum crashen. In der Schleife die ausgeführt wird stimmt was nicht mit der Array (bzw. Dict) Indizierung

Reply:
(In meinem Test Experiment wurde Bias benutzt, wird wahrscheinlich daran liegen)

Reply:
Ja, Bias führt zum Abbruch.

Reply:
Ja, das mit dem Bias funktioniert nicht. Wir hatten uns darauf geeinigt, dass ich es nicht umsetzen werde.